roles/test-utils/mining-device/src/main.rs

```diff
@@ -669,3 +686,74 @@ fn generate_random_32_byte_array() -> [u8; 32] {
     rng.fill(&mut arr[..]);
     arr
 }
+
+fn start_mining_threads(
+    have_new_job: Receiver<()>,
+    miner: Arc<Mutex<Miner>>,
+    share_send: Sender<(u32, u32, u32, u32)>,
+) {
+    tokio::task::spawn(async move {
+        let mut killers: Vec<Arc<AtomicBool>> = vec![];
+        loop {
+            let available_parallelism = u32::max(
+                2,
+                std::thread::available_parallelism().unwrap().get() as u32,
+            );
+            let p = available_parallelism - 1;
+            let unit = u32::MAX / p;
+            while have_new_job.recv().await.is_ok() {
+                while let Some(killer) = killers.pop() {
+                    killer.store(true, Ordering::Relaxed);
+                }
+                let miner = miner.safe_lock(|m| m.clone()).unwrap();
+                for i in 0..p {
+                    let mut miner = miner.clone();
+                    let share_send = share_send.clone();
+                    let killer = Arc::new(AtomicBool::new(false));
+                    miner.header.as_mut().map(|h| h.nonce = i * unit);
+                    killers.push(killer.clone());
+                    std::thread::spawn(move || {
+                        mine(miner, share_send, killer);
+                    });
+                }
+            }
+        }
+    });
+}
+
+fn mine(mut miner: Miner, share_send: Sender<(u32, u32, u32, u32)>, kill: Arc<AtomicBool>) {
+    if miner.handicap != 0 {
+        loop {
+            if kill.load(Ordering::Relaxed) {
+                break;
+            }
+            std::thread::sleep(std::time::Duration::from_micros(miner.handicap.into()));
+            if miner.next_share().is_valid() {
+                let nonce = miner.header.unwrap().nonce;
+                let time = miner.header.unwrap().time;
+                let job_id = miner.job_id.unwrap();
+                let version = miner.version;
+                share_send
+                    .try_send((nonce, job_id, version.unwrap(), time))
+                    .unwrap();
+            }
+            miner.header.as_mut().map(|h| h.nonce += 1);
+        }
+    } else {
+        loop {
+            if miner.next_share().is_valid() {
+                if kill.load(Ordering::Relaxed) {
+                    break;
+                }
+                let nonce = miner.header.unwrap().nonce;
+                let time = miner.header.unwrap().time;
+                let job_id = miner.job_id.unwrap();
+                let version = miner.version;
+                share_send
+                    .try_send((nonce, job_id, version.unwrap(), time))
+                    .unwrap();
+            }
+            miner.header.as_mut().map(|h| h.nonce += 1);
+        }
+    }
+}
```

Comment:
we can remove some redundancies in the logic of this function, there's no need for two separate loops:
```suggestion
fn mine(mut miner: Miner, share_send: Sender<(u32, u32, u32, u32)>, kill: Arc<AtomicBool>) {
    loop {
        if kill.load(Ordering::Relaxed) {
            break;
        }
        if miner.handicap > 0 {
            std::thread::sleep(std::time::Duration::from_micros(miner.handicap.into()));
        }
        if miner.next_share().is_valid() {
            let nonce = miner.header.unwrap().nonce;
            let time = miner.header.unwrap().time;
            let job_id = miner.job_id.unwrap();
            let version = miner.version;
            share_send
                .try_send((nonce, job_id, version.unwrap(), time))
                .unwrap();
        }
        miner.header.as_mut().map(|h| h.nonce += 1);
    }
}
```

Reply:
as already stated above this is very hot code and I prefer to keep it like that.